hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServletInodeBasedXfer.java

```diff
@@ -0,0 +1,386 @@
+      response.setContentType("application/x-tar");
+      response.setHeader("Content-Disposition", "attachment; filename=\"" + tarName + "\"");
+      Instant start = Instant.now();
+      writeDbDataToStream(request, response.getOutputStream(), receivedSstFiles, tmpdir);
+      Instant end = Instant.now();
+      long duration = Duration.between(start, end).toMillis();
+      LOG.info("Time taken to write the checkpoint to response output " +
+          "stream: {} milliseconds", duration);
+      logSstFileList(excludedSstList,
+          "Excluded {} SST files from the latest checkpoint{}: {}", 5);
+    } catch (Exception e) {
+      LOG.error(
+          "Unable to process metadata snapshot request. ", e);
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    } finally {
+      try {
+        if (tmpdir != null) {
+          FileUtils.deleteDirectory(tmpdir.toFile());
+        }
+      } catch (IOException e) {
+        LOG.error("unable to delete: " + tmpdir, e.toString());
+      }
+    }
+  }
+
+  Path getSstBackupDir() {
+    RocksDBCheckpointDiffer differ = getDbStore().getRocksDBCheckpointDiffer();
+    return new File(differ.getSSTBackupDir()).toPath();
+  }
+
+  Path getCompactionLogDir() {
+    RocksDBCheckpointDiffer differ = getDbStore().getRocksDBCheckpointDiffer();
+    return new File(differ.getCompactionLogDir()).toPath();
+  }
+
+  public void writeDbDataToStream(HttpServletRequest request, OutputStream destination,
+      Set<String> sstFilesToExclude, Path tmpdir) throws IOException {
+    DBCheckpoint checkpoint = null;
+    OzoneManager om = (OzoneManager) getServletContext().getAttribute(OzoneConsts.OM_CONTEXT_ATTRIBUTE);
+    OMMetadataManager omMetadataManager = om.getMetadataManager();
+    boolean includeSnapshotData = includeSnapshotData(request);
+    AtomicLong maxTotalSstSize = new AtomicLong(getConf().getLong(OZONE_OM_RATIS_SNAPSHOT_MAX_TOTAL_SST_SIZE_KEY,
+        OZONE_OM_RATIS_SNAPSHOT_MAX_TOTAL_SST_SIZE_DEFAULT));
+
+    Set<Path> snapshotPaths = Collections.emptySet();
+
+    if (!includeSnapshotData) {
+      maxTotalSstSize.set(Long.MAX_VALUE);
+    } else {
+      snapshotPaths = getSnapshotDirs(omMetadataManager);
+    }
+
+    if (sstFilesToExclude.isEmpty()) {
+      logEstimatedTarballSize(getDbStore().getDbLocation().toPath(), snapshotPaths);
+    }
+
+    boolean shouldContinue = true;
+
+    Map<String, String> hardLinkFileMap = new HashMap<>();
+    try (ArchiveOutputStream<TarArchiveEntry> archiveOutputStream = tar(destination)) {
+      if (includeSnapshotData) {
+        // Process each snapshot db path and write it to archive
+        for (Path snapshotDbPath : snapshotPaths) {
+          if (!shouldContinue) {
+            break;
+          }
+          shouldContinue = writeDBToArchive(sstFilesToExclude, snapshotDbPath,
+              maxTotalSstSize, archiveOutputStream, tmpdir, hardLinkFileMap);
```

Comment:
Should we also include `ratisSnapshotComplete` file for each snapshot transfer that is completed? 
That way we can add another section to the form data called `OZONE_DB_CHECKPOINT_REQUEST_TO_EXCLUDE_SNAPSHOTS` [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis_snapshot/OmRatisSnapshotProvider.java#L230-L231) and just skip processing the snapshot in subsequent requests.
And `OZONE_DB_CHECKPOINT_REQUEST_TO_EXCLUDE_SST_FILES` will only have files from the ongoing snapshot transfer.

Reply:
No , we cannot do this as snapshot DB's  can change example : while moving keys from one snapshot to another during snapshot delete.